server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/BaseStarTreeBuilder.java

```diff
@@ -155,6 +174,228 @@ public List<MetricAggregatorInfo> generateMetricAggregatorInfos(MapperService ma
+            appendToStarTree(starTreeDocumentIterator.next());
+        }
+    }
+
+    private void serializeStarTree(int numSegmentStarTreeDocument, int numStarTreeDocs) throws IOException {
+        // serialize the star tree data
+        long dataFilePointer = dataOut.getFilePointer();
+        StarTreeWriter starTreeWriter = new StarTreeWriter();
+        long totalStarTreeDataLength = starTreeWriter.writeStarTree(dataOut, rootNode, numStarTreeNodes, starTreeField.getName());
+
+        // serialize the star tree meta
+        starTreeWriter.writeStarTreeMetadata(
+            metaOut,
+            starTreeField,
+            metricAggregatorInfos,
+            numStarTreeNodes,
+            numSegmentStarTreeDocument,
+            numStarTreeDocs,
+            dataFilePointer,
+            totalStarTreeDataLength
+        );
+    }
+
+    private void createSortedDocValuesIndices(DocValuesConsumer docValuesConsumer, AtomicInteger fieldNumberAcrossStarTrees)
+        throws IOException {
+        List<SortedNumericDocValuesWriterWrapper> dimensionWriters = new ArrayList<>();
+        List<SortedNumericDocValuesWriterWrapper> metricWriters = new ArrayList<>();
+        FieldInfo[] dimensionFieldInfoList = new FieldInfo[starTreeField.getDimensionsOrder().size()];
+        FieldInfo[] metricFieldInfoList = new FieldInfo[metricAggregatorInfos.size()];
+        for (int i = 0; i < dimensionFieldInfoList.length; i++) {
+            final FieldInfo fi = getFieldInfo(
+                fullyQualifiedFieldNameForStarTreeDimensionsDocValues(
+                    starTreeField.getName(),
+                    starTreeField.getDimensionsOrder().get(i).getField()
+                ),
+                DocValuesType.SORTED_NUMERIC,
+                fieldNumberAcrossStarTrees.getAndIncrement()
+            );
+            dimensionFieldInfoList[i] = fi;
+            dimensionWriters.add(new SortedNumericDocValuesWriterWrapper(fi, Counter.newCounter()));
+        }
+        for (int i = 0; i < metricAggregatorInfos.size(); i++) {
+
+            final FieldInfo fi = getFieldInfo(
+                fullyQualifiedFieldNameForStarTreeMetricsDocValues(
+                    starTreeField.getName(),
+                    metricAggregatorInfos.get(i).getField(),
+                    metricAggregatorInfos.get(i).getMetricStat().getTypeName()
+                ),
+                DocValuesType.SORTED_NUMERIC,
+                fieldNumberAcrossStarTrees.getAndIncrement()
+            );
+
+            metricFieldInfoList[i] = fi;
+            metricWriters.add(new SortedNumericDocValuesWriterWrapper(fi, Counter.newCounter()));
+        }
+
+        for (int docId = 0; docId < numStarTreeDocs; docId++) {
+            StarTreeDocument starTreeDocument = getStarTreeDocumentForCreatingDocValues(docId);
+            for (int i = 0; i < starTreeDocument.dimensions.length; i++) {
+                Long val = starTreeDocument.dimensions[i];
```

Comment:
why not 
```
if(starTreeDocument.dimensions[i] != null) {
    dimensionWriters.get(i).addValue(docId, starTreeDocument.dimensions[i]);
}
```

since this is called for every doc , lets be as optimized as possible